### ISSUE TITLE
Tags invalidation not working 2 (Symfony 5.4)

### ISSUE THREAD

**Issue by ii02735** [FriendsOfSymfony/FOSHttpCacheBundle#596]
Hello,

I have the same issue when I try to invalidate a tag with the CLI, in a Symfony 5.4 project, with Symfony HttpCache.

Here is my configuration file : 

```yml
fos_http_cache:
  debug:
    enabled: true
  proxy_client:
    symfony:
      use_kernel_dispatcher: true
  cache_control:
    defaults:
      overwrite: false
    rules:
      -
        match:
          attributes: 
            _route: ^api_abilities|api_ability_by_id$
        headers:
          cache_control: { public: true, s_maxage: 2629800 }
  tags:
    enabled: true
    rules:
      - # ability 
        match:
          attributes: 
            _route: ^api_abilities$
        tags: ['abilities']

      -
        match:
          attributes: 
            _route: ^api_abilities_by_id$
        tags: ['abilities']
        tag_expressions: ["'ability-'~id"]
```

My CacheKernel class file : 

```php
class CacheKernel extends HttpCache implements CacheInvalidation
{
    use EventDispatchingHttpCache;

    /**
     * Overwrite constructor to register event listeners for FOSHttpCache.
     *
     * @param SurrogateInterface $surrogate
     * @param mixed[] $options
     */
    public function __construct(
        Kernel $kernel,
        SurrogateInterface $surrogate = null,
        array $options = []
    ) {
        $store = new Psr6Store([
            'cache_directory' => $kernel->getCacheDir()
        ]);
        parent::__construct($kernel, $store, $surrogate, $options);

        $this->addSubscriber(new PurgeListener());
        $this->addSubscriber(new RefreshListener());
        $this->addSubscriber(new UserContextListener());
        $this->addSubscriber(new PurgeTagsListener());
        $this->addSubscriber(new CleanupCacheTagsListener());
        if (isset($options['debug']) && $options['debug']) {
            $this->addSubscriber(new DebugListener());
        }
    }

    /**
     * Made public to allow event listeners to do refresh operations.
     *
     * {@inheritDoc}
     */
    public function fetch(Request $request, $catch = false)
    {
        return parent::fetch($request, $catch);
    }
}
```

And my Kernel class file : 

```php
class Kernel extends BaseKernel implements HttpCacheProvider
{
    use HttpCacheAware;
    use MicroKernelTrait;

    public function __construct(string $environment, bool $debug)
    {
        parent::__construct($environment, $debug);
        $this->setHttpCache(new CacheKernel($this));
    }
```

**The steps of my test :**

1. I call my `api_abilities` route (`/api/abilities`), and the response headers are the following : 

```http
Server: nginx/1.21.4
Content-Type: application/json
Content-Length: 89512
Connection: keep-alive
X-Powered-By: PHP/7.4.33
Set-Cookie: XDEBUG_SESSION=PHPSTORM; path=/; SameSite=Lax
Cache-Control: public, s-maxage=2629800
Date: Fri, 01 Sep 2023 16:14:18 GMT
X-Cache-Debug: 1
X-Cache-Tags: abilities
X-Content-Digest: en74cdf57df9307601a67fad384500a6fc3b1dd4be3a3f56ad1f7d74b2212be4c8
Age: 0
```

The `Age` is 0, which make sense, because it is the first time that I called this route.
If i make a second call of the same route, the `Age` header has increased, so the content has been clearly cached.

2. I make a change in my database in order to make the `api_abilities` response invalid/stale.
So, I run the following command : `bin/console fos:httpcache:invalidate:tag abilities`, the verbose output is : 

```
Sent 1 invalidation request(s)
```
3. I make a new call to `api_abilities`.

**What I expect :** a response with fresh data

**What I got :** the same data as the previous request so the date is still the same as above, and the age didn't reset to 0.

```http
Server: nginx/1.21.4
Content-Type: application/json
Content-Length: 89512
Connection: keep-alive
X-Powered-By: PHP/7.4.33
Set-Cookie: XDEBUG_SESSION=PHPSTORM; path=/; SameSite=Lax
Cache-Control: public, s-maxage=2629800
date: Fri, 01 Sep 2023 16:14:18 GMT
x-cache-debug: 1
x-cache-tags: abilities
x-content-digest: en74cdf57df9307601a67fad384500a6fc3b1dd4be3a3f56ad1f7d74b2212be4c8
Age: 478
```

Am I missing something ?

_Originally posted by @ii02735 in https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/issues/501#issuecomment-1703015451_

**Comment by ii02735** [comment]
> i don't see whats wrong. can you set a breakpoint in https://github.com/Toflar/psr6-symfony-http-cache-store/blob/25623d8e0858223f6849d2200d36a43df17c655d/src/Psr6Store.php#L275 to validate if the invalidation on the store is actually called?
> 
@dbu I set a breakpoint at the location you specified, ~but it didn't reach it when I a run the command. It did reach however the `invalidate(Request $request): void` of Psr6Store.php.~ it reached the method as expected :thinking: 

Then it reached this method from the `$this->cache->invalidateTags($tags);` line : 

```php
# vendor/symfony/cache/Adapter/AbstractTagAwareAdapter.php
    /**
     * {@inheritdoc}
     */
    public function invalidateTags(array $tags)
    {
        if (empty($tags)) {
            return false;
        }

        $tagIds = [];
        foreach (array_unique($tags) as $tag) {
            $tagIds[] = $this->getId(self::TAGS_PREFIX.$tag);
        }

        try {
            if ($this->doInvalidate($tagIds)) {
                return true;
            }
        } catch (\Exception $e) {
            CacheItem::log($this->logger, 'Failed to invalidate tags: '.$e->getMessage(), ['exception' => $e, 'cache-adapter' => get_debug_type($this)]);
        }

        return false;
    }
```

**Comment by ii02735** [comment]
Oh, it seems that I have a grasp of the issue there.

When the `doInvalidate` method is triggered, this following condition fails : 

```php
# vendor/symfony/cache/Adapter/FilesystemTagAwareAdapter.php
    /**
     * {@inheritdoc}
     */
    protected function doInvalidate(array $tagIds): bool
    {
        foreach ($tagIds as $tagId) {
            if (!is_dir($tagFolder = $this->getTagFolder($tagId))) {
                continue;
            }
```
The desired directory called `tags` does not exist :thinking: (`var/cache/prod/@/tags/G/N/N3OB8z6fpSbBXrnpPTgA`)

**Comment by ii02735** [comment]
I found my mistake, my apologies : It seems that I didn't provide the correct path for Psr6Store.

I added `/http_cache` for the `cache_directory` option, and it worked :

```php

    public function __construct(
        Kernel $kernel,
        SurrogateInterface $surrogate = null,
        array $options = []
    ) {
        $store = new Psr6Store([
            'cache_directory' => $kernel->getCacheDir().'/http_cache',
            'cache_tags_header' => self::TAGS_HEADER
        ]);
```

**Comment by dbu** [comment]
ah, glad you found the issue, and thanks for reporting back!

i am however a bit confused why you need to add a path to the kernel cache dir. in [our documentation](https://foshttpcache.readthedocs.io/en/latest/symfony-cache-configuration.html#tagging) we just say `$kernel->getCacheDir()`. 

and tag invalidation should go through the same instance of that psr6store, so the path should be allowed to be whatever it wants.

@Toflar do you see what is wrong here?

**Comment by Toflar** [comment]
No, it should work out of the box. Maybe the responses weren't correctly tagged in the first place so the tags directory wasn't added? Should be easy to check: The `tags` folder should contain the info upon the first response :) Might be a bug in Symfony but so far I haven't experienced any problems myself - this would be pretty surprising.

**Comment by ii02735** [comment]
Hello @dbu and @Toflar, you're both right, the mistake was mine :pray: 

After reviewing again my configuration, I noticed two things :

1. I made a mistake in the `_route` attribute for the second rule of the `tags` section : I wrote `^api_abilities_by_id$` instead of `^api_ability_by_id$`...

2. There was a mismatch for the cache tags header between Psr6Store and the PurgeListener : the former uses `Cache-Tags` by default, and the latter, `X-Cache-Tags`. I had to change the value of my `TAGS_HEADER` constant to ensure that it works properly.

I didn't actually need to change the `cache_directory` option :sweat_smile: 

Thank you for your work and your time !
